### ISSUE TITLE
Views referencing each other can't be created simultaneously

### ISSUE THREAD

**Issue by VictorLeP** [olirice/alembic_utils#9]
I can't seem to create multiple views, where one references a previous view, at the same time.

An minimal example of what I would like to define (at the same time, so both views should be generated within one migration):
```python
first_view = PGView(
    schema="public",
    signature="first_view",
    definition="SELECT * FROM some_table;"
)
second_view = PGView(
    schema="public",
    signature="second_view",
    definition="SELECT * FROM first_view;"
)
```

I get the following error while trying to autogenerate the alembic migration: `psycopg2.errors.UndefinedTable: relation "second_view" does not exist`.
It seems that while comparing the existing views with the newly defined views, alembic actually attempts to execute the view definition, which fails for the second view as the first view is then not yet defined.
https://github.com/olirice/alembic_utils/blob/a4bc7f5f025335faad7b178eb84ab78093e525ec/src/alembic_utils/replaceable_entity.py#L28-L40

**Comment by olirice** [comment]
Are you producing the migrations manually, or using `--autogenerate`?

When using autogenerate there would be a `op.create_entity(first_view)` call between the declaration of `first_view` and `second_view` which should solve the dependency problem.

i.e.
```python3
first_view = PGView(
    schema="public",
    signature="first_view",
    definition="SELECT * FROM some_table;"
)
op.create_entity(first_view)

second_view = PGView(
    schema="public",
    signature="second_view",
    definition="SELECT * FROM first_view;"
)
op.create_entity(first_view)
```

You may need to re-order the view definitions so the dependencies of each create statement exist when they're executed. Your example doesn't have that problem, but good to be aware of

If that doesn't answer the question please post a full migration file and versions for python, alembic, and alembic_utils so I can try to reproduce

**Comment by VictorLeP** [comment]
I'm using `--autogenerate`. I think the issue occurs during the autogeneration (i.e. reading the `PGView`s registered with `register_entities`): this fails already and alembic doesn't reach the stage of generating `op.create_entity`. In fact, manually creating the migration file (similar to your code) actually works fine.

**Comment by VictorLeP** [comment]
Versions: 
* python 3.8.0
* alembic 1.4.2
* alembic-utils 0.2.7

**Comment by olirice** [comment]
Good point, I didn't consider the rendering step.

At the moment, the only solution would be to manually write the migration as shown ^, or autogenerate a new migration when you want to checkpoint and be able to refer to the new entities like:
- add `first_view` to source
- autogenerate migration for `first_view`
- apply the migration
- add `second_view` to source
- autogenerate migration for `second_view`

**Comment by VictorLeP** [comment]
Indeed, the first solution seems more convenient (most of it is copy-paste from the original definition), the second one gets more inconvenient for every additional view you want to define at the same time.

**Comment by Woodz** [comment]
As far as I can see, this problem occurs because all defined SQL views are created in a dummy schema `alembic_utils` to allow comparison between the current versions of the SQL views (https://github.com/olirice/alembic_utils/blob/65d8a07fe0f8764299f9e80cb95262aa151ec9a2/src/alembic_utils/replaceable_entity.py#L27). This means that the second view is trying to reference the first view in the normal schema, when it only exists in the dummy schema at this point. One option would be to replace the schema in the second view when creating in the dummy schema, but I'm unsure how tricky this would be to implement.

**Comment by olirice** [comment]
There are a bunch of issues:
- Replacing the schema name in the second view is a good thought but it would be very difficult given that we don't parse entity definitions in python.
- Entities are "simulated" in the `alembic_utils` schema sequentially and 1 at a time so no dependency links are valid among uncommitted entities
- Assuming we solve the previous two problems, we still need a dependency resolver to sort the migration operations in an order that will succeed


I wrote a draft this morning 
- swaps the `alembic_utils` dummy schema for transactional ddl
- can simulate multiple entities simultaneously
- resolves dependencies

I'll ping you on it for review at some point today

**Comment by olirice** [comment]
Resolved in `alembic_utils>=0.3.2`

Tested [here](https://github.com/olirice/alembic_utils/blob/master/src/test/test_dependency_resolver.py)
Not all cases are covered but general use is expected to work

If you end up on this thread and are running into an problem please open a new issue with minimal reproducible example.
